src/jvm/main/org/jetbrains/kotlinx/lincheck/util/Spinner.kt

```diff
@@ -0,0 +1,259 @@
+     * then the spinner should exit the loop immediately.
+     */
+    private val shouldSpin: Boolean = run {
+        val nProcessors = Runtime.getRuntime().availableProcessors()
+        (nProcessors >= nThreads)
+    }
+
+    /**
+     * The number of spin-loop iterations to be performed per call to [spin].
+     */
+    private val spinLoopIterationsPerCall: Int =
+        if (shouldSpin) SPIN_LOOP_ITERATIONS_PER_CALL else 1
+
+    /**
+     * The number of spin-loop iterations before yielding the current thread
+     * to give other threads the opportunity to run.
+     */
+    private val yieldLimit = when {
+        shouldYield && shouldSpin -> SPIN_LOOP_ITERATIONS_BEFORE_YIELD
+        shouldYield -> 1
+        else -> -1
+    }
+
+    /**
+     * Determines whether the spinner should yield to other threads based on the yield limit.
+     */
+    private inline val shouldYield: Boolean
+        get() = yieldLimit > 0
+
+    /**
+     * The exit limit determines the number of spin-loop iterations
+     * after which the spin-loop is advised to exit.
+     */
+    private val exitLimit = when {
+        shouldSpin -> SPIN_LOOP_ITERATIONS_BEFORE_EXIT
+        else -> 1
+    }
+
+    /**
+     * Spins the counter for a few iterations.
+     * In addition, in the case of a yielding spinner, it yields occasionally
+     * to give other threads the opportunity to run.
+     *
+     * @return `true` if the spin-loop should continue;
+     *   `false` if the spin-loop is advised to exit,
+     *   for example, to fall back into a blocking synchronization.
+     */
+    fun spin(): Boolean {
+        // spin a few iterations
+        repeat(spinLoopIterationsPerCall) {
+            counter++
+        }
+        // if yield limit is approached,
+        // then yield and give other threads the opportunity to run
+        if (shouldYield && counter % yieldLimit == 0) {
+            Thread.yield()
+        }
+        // if exit limit is approached,
+        // reset counter and signal to exit the spin-loop
+        if (counter >= exitLimit) {
+            counter = 0
+            return false
+        }
+        return true
+    }
+
+    /**
+     * Resets the state of the spinner.
+     */
+    fun reset() {
+        counter = 0
+    }
+
+}
+
+/**
+ * A [SpinnerGroup] represents a group of spinners that can be accessed by their index.
+ * It provides a convenient way to manage multiple spinners together.
+ */
+@JvmInline
+value class SpinnerGroup private constructor(private val spinners: Array<Spinner>) {
+
+    /**
+     * This constructor creates a spinners group to be used by the specified number of threads.
+     *
+     * @param nThreads The number of threads in the group.
+     * @param shouldYield Determines whether the spin-loop in each Spinner should periodically yield
+     *   to give other threads the opportunity to run.
+     */
+    constructor(nThreads: Int, shouldYield: Boolean = true)
```

Comment:
If I'm not mistaken, `shouldYield` is always `true`

Reply:
Yes, it looks like so. 
So should I remove this parameter?

Reply:
I removed it.

Reply:
Thanks!